### PR TITLE
Adopt new regex.yaml format (with regex flag, device band and model)

### PR DESCRIPTION
**Pull request by alonbg**
Hi there,

Two issues not yet solved:
1. There are placeholder leftovers after replacing (just for device.* properties).
2. Device tests passed: 15318/15948  uap-core/tests/test_device.yaml.

I have tried regex and regex_search flags and compared to the uap-python impl (which seem to pass all tests)
If you have any ideas/directions ...

Anyway, It would be great if you could merge.

cheers

### CODE REVIEW

File: UaParser.cpp [hunk 1]
```diff
@@ -6,20 +6,27 @@
...
 #include <glog/logging.h>
 #include <yaml-cpp/yaml.h>
 
 namespace {
 
-struct DeviceStore {
-  std::string replacement;
-  boost::regex regExpr;
-};
+  struct GenericStore {
```
Comment (asuhan) [inline UaParser.cpp]: Restore indentation here.

File: UaParser.cpp [hunk 2]
```diff
@@ -6,20 +6,27 @@
...
+      std::string replacement;
+      boost::regex regExpr;
+  };
 
-struct AgentStore : DeviceStore {
-  std::string majorVersionReplacement;
-  std::string minorVersionReplacement;
-};
+  struct DeviceStore : GenericStore {
+    std::string brand_replacement;
```
Comment (asuhan) [inline UaParser.cpp]: Camel-case the name of these for consistency.

File: UaParser.cpp [hunk 3]
```diff
@@ -87,6 +102,36 @@ struct UAStore {
...
 
+
+////////////
+// HELPERS //
+/////////////
+
+//Device parsing is different than the other parsing in that a field can have many placeholders
+const char* placeholders[]= {"$1", "$2", "$3", "$4","$5", "$6", "$7", "$8", "$9"};
+void replace_all_placeholders(std::string& device_property, const boost::smatch &result) {
+    std::string::size_type loc;
```
Comment (asuhan) [inline UaParser.cpp]: Use 2 space indentation here and generally make it consistent with the rest. `idx` in line 114 should be `size_t`. Remove the spaces before / after parentheses in lines 117, 118 127 and add spaces after comas.

File: main.cpp [hunk 4]
```diff
@@ -11,14 +11,14 @@
 # distributed under the License is distributed on an 'AS IS' BASIS,
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
-# limitations under the License.
+# limitations under the License.#include "ua_parser.h"
```
Comment (asuhan) [inline main.cpp]: Revert the accidental changes here,

File: UaParser.cpp [hunk 5]
```diff
@@ -87,6 +102,36 @@ struct UAStore {
...
+            if ( result[idx].matched ) {
+              device_property.replace(loc,2,result[idx].str());
+            }
+            else
+              device_property.erase(loc,2);
+        }
+    }
+
+    // There should not be placehoders leftover. This is a Workaround ...
+    loc = device_property.find_first_of('$');
```
Comment (asuhan) [inline UaParser.cpp]: Does this actually happen in tests? Looks safe though. Also, remove additional new lines.

**Comment by asuhan** [comment]
@alonbg Thanks for doing this, the pull request looks great modulo a few nits you can see inline. Once you fix them I'll take another, thorough look by the end of the week and we should merge this to upstream ASAP.

File: UaParser.cpp (hunk 5 above)
Comment (alonbg) [inline UaParser.cpp]: yep, this is one of the two issues I mentioned.

The second issue is regex_search (somehow) misses the specific patterns and eventually hits the generic patterns at the end. I changed ASSERT_EQ to EXPECT_EQ for the devices test until a solution is found.

**Pull request closed without merging**